sources/host-ctr/cmd/host-ctr/main.go

```diff
@@ -575,16 +577,102 @@ func cleanUp(containerdSocket string, namespace string, containerID string) erro
+	}
+	account := matches[1]
+	region := matches[2]
+
+	// Need to include the full repository path and the imageID (e.g. /eks/image-name:tag)
+	tokens := strings.SplitN(input, "/", 2)
+	if len(tokens) != 2 {
+		return ecr.ECRSpec{}, fmt.Errorf("invalid image URI: %s", input)
+	}
+	fullRepoPath := tokens[len(tokens)-1]
+	// Run simple checks on the provided repository.
+	switch {
+	case
+		// Must not be empty
+		fullRepoPath == "",
+		// Must not have a partial/unsupplied label
+		strings.HasSuffix(fullRepoPath, ":"),
+		// Must not have a partial/unsupplied digest specifier
+		strings.HasSuffix(fullRepoPath, "@"):
+		return ecr.ECRSpec{}, errors.New("incomplete reference provided")
+	}
+
+	// Parse out image reference's to validate.
+	ref, err := reference.Parse("repository/" + fullRepoPath)
+	if err != nil {
+		return ecr.ECRSpec{}, err
+	}
+	// If the digest is provided, check that it is valid.
+	if ref.Digest() != "" {
+		err := ref.Digest().Validate()
+		// Digest may not be supported by the client despite it passing against
+		// a rudimentary check. The error is different in the passing case, so
+		// that's considered a passing check for unavailable digesters.
+		//
+		// https://github.com/opencontainers/go-digest/blob/ea51bea511f75cfa3ef6098cc253c5c3609b037a/digest.go#L110-L115
+		if err != nil && err != digest.ErrDigestUnsupported {
+			return ecr.ECRSpec{}, fmt.Errorf("%v: %w", "invalid image URI", err)
+		}
+	}
+
+	// Get the ECR image reference prefix from the AWS region
+	ecrRefPrefix, ok := ecrRefPrefixMapping[region]
+	if !ok {
+		return ecr.ECRSpec{}, fmt.Errorf("%s: %s", "invalid region in internal mapping", region)
+	}
+
+	return ecr.ParseRef(fmt.Sprintf("%s%s:repository/%s", ecrRefPrefix, account, fullRepoPath))
+}
+
+// fetchECRRef attempts to resolve the ECR reference from an input source string
+// by first using the aws-sdk-go's ParseImageURI function. This will fail for
+// special regions that are not yet supported. If it fails for any reason,
+// attempt to parse again using parseImageURISpecialRegions in this package.
+// This uses a special region reference to build the ECR image references.
+// If both fail, an error is returned.
+func fetchECRRef(ctx context.Context, input string) (ecr.ECRSpec, error) {
+	var spec ecr.ECRSpec
+	spec, err := ecr.ParseImageURI(input)
+	if err == nil {
+		return spec, nil
+	}
+	log.G(ctx).WithError(err).WithField("source", input).Error("failed to parse ECR reference")
+
+	// The parsing might fail if the AWS region is special, parse again with special handling:
+	spec, err = parseImageURISpecialRegions(input)
+	if err != nil {
+		log.G(ctx).WithError(err).WithField("source", input).Error("failed to parse special ECR reference")
+
+		// Return the original error and the new error generated by parsing
+		// for the special region
+		return ecr.ECRSpec{}, errors.Wrap(err, "could not parse ECR reference for special regions")
+	}
+
+	return spec, nil
```

Comment:
nit: can early return here too for symmetry with the above case --
```suggestion
	// The parsing might fail if the AWS region is special, parse again with special handling:
	spec, err = parseImageURISpecialRegions(input)
	if err == nil {
		return spec, nil
	}
	
	log.G(ctx).WithError(err).WithField("source", input).Error("failed to parse special ECR reference")
	
	// Return the original error and the new error generated by parsing
	// for the special region
	return ecr.ECRSpec{}, errors.Wrap(err, "could not parse ECR reference for special regions")
```